bootstrap/cmd/bootstrap/app/ksServer.go

```diff
@@ -527,18 +494,75 @@ func (s *ksServer) autoConfigureApp(kfApp *kApp.App, appConfig *AppConfig, names
 	return nil
 }
 
-func (s *ksServer) GetApp(project string, appName string) (*appInfo, error) {
-	mapKey := fmt.Sprintf(" %s-%s", project, appName)
-	info, ok := s.apps[mapKey]
-	if !ok {
+
+func GetRepoName(project string) string {
+	return fmt.Sprintf("%s-kubeflow-config", project)
+}
+
+// Not thread-safe, make sure project lock is on.
+func (s *ksServer) CloneRepoToLocal(project string, token string) error {
+	repoDir := path.Join(s.appsDir, GetRepoName(project))
+	_, err := s.fs.Stat(repoDir)
+
+	// delete existing local repo, pull from cloud
+	if err == nil {
+		os.RemoveAll(repoDir)
+	}
+	ts := oauth2.StaticTokenSource(&oauth2.Token{
+		AccessToken: token,
+	})
+	sourcerepoService, err := sourcerepo.New(oauth2.NewClient(context.Background(), ts))
+	_, err = sourcerepoService.Projects.Repos.Get(fmt.Sprintf("projects/%s/repos/%s", project, GetRepoName(project))).Do()
+	if err != nil {
+		// repo does't exist in target project, create one
+		_, err = sourcerepoService.Projects.Repos.Create(fmt.Sprintf("projects/%s", project), &sourcerepo.Repo{
+			Name: fmt.Sprintf("projects/%s/repos/%s", project, GetRepoName(project)),
+		}).Do()
+		if err != nil {
+			return err
+		}
+	}
+	err = os.Chdir(s.appsDir)
+	if err != nil {
+		return err
+	}
+	cloneCmd := fmt.Sprintf("git clone https://%s:%s@source.developers.google.com/p/%s/r/%s",
+		"user1", token, project, GetRepoName(project))
+	cmd := exec.Command("sh", "-c", cloneCmd)
+	result, err := cmd.CombinedOutput()
+	log.Infof(string(result))
+	return err
+}
+
+func (s *ksServer) GetApp(project string, appName string, token string) (*appInfo, error) {
+	s.CloneRepoToLocal(project, token)
+	appDir := path.Join(s.appsDir, GetRepoName(project), appName)
+	_, err := s.fs.Stat(appDir)
+	if err != nil {
 		return nil, fmt.Errorf("App %s doesn't exist in Project %s", appName, project)
 	}
-	return info, nil
+	kfApp, err := kApp.Load(s.fs, appDir, true)
+
+	if err != nil {
+		return nil, fmt.Errorf("There was a problem loading app %v. Error: %v", appName, err)
+	}
+
+	return &appInfo{
+		App: kfApp,
+	}, nil
 }
+
+// Save ks app config to cloud source repos
 // Not thread safe, be aware when call it.
-func (s *ksServer) InsertApp(project string, appName string, app *appInfo) {
-	mapKey := fmt.Sprintf(" %s-%s", project, appName)
-	s.apps[mapKey] = app
+func (s *ksServer) SaveAppToRepo(project string) error {
```

Comment:
Can you add some comment on this. It will sync all the app dir of the project right?

Reply:
It will sync the project repo, and each project has one repo.